classification/scripts/models.py

```diff
@@ -0,0 +1,191 @@
+# --------------------
+
+class InitialConv(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super().__init__()
+        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=7,
+                              stride=2, padding=3)
+        self.bn = nn.BatchNorm2d(out_channels)
+        self.relu = nn.ReLU(inplace=True)
+
+    def forward(self, x):
+        return self.relu(self.bn(self.conv(x)))
+
+
+class MaskAttention(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, img_feat, mask_feat):
+        if mask_feat.shape[2:] != img_feat.shape[2:]:
+            mask_feat = F.interpolate(mask_feat, size=img_feat.shape[2:],
+                                      mode='bilinear', align_corners=False)
+        attention = torch.sigmoid(mask_feat)
+        return img_feat * attention
+
+
+class ResBlock(nn.Module):
+    def __init__(self, in_channels, out_channels, stride=1):
+        super().__init__()
+        self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size=3,
+                               stride=stride, padding=1)
+        self.bn1 = nn.BatchNorm2d(out_channels)
+        self.relu = nn.ReLU(inplace=True)
+
+        self.conv2 = nn.Conv2d(out_channels, out_channels, kernel_size=3,
+                               padding=1)
+        self.bn2 = nn.BatchNorm2d(out_channels)
+
+        self.skip = nn.Sequential()
+        if in_channels != out_channels or stride != 1:
+            self.skip = nn.Sequential(
+                nn.Conv2d(in_channels, out_channels, kernel_size=1,
+                          stride=stride),
+                nn.BatchNorm2d(out_channels)
+            )
+
+    def forward(self, x):
+        identity = self.skip(x)
+        out = self.relu(self.bn1(self.conv1(x)))
+        out = self.bn2(self.conv2(out))
+        out += identity
+        return self.relu(out)
+
+
+# --------------------
+# ResNet Siamese Network
+# --------------------
+
+def get_pretrained_resnet34(output_size, freeze=False):
+    model_conv = models.resnet34(pretrained=True)
+
+    for param in model_conv.parameters():
+        param.requires_grad = False
+
+    num_ftrs = model_conv.fc.in_features
+    model_conv.fc = nn.Linear(num_ftrs, output_size)
+
+    if not freeze:
+        for name, child in model_conv.named_children():
+            if name in ["layer3", "layer4"]:
+                for param in child.parameters():
+                    param.requires_grad = True
+
+    return model_conv
+
+
+class ResNetSiameseNetwork(nn.Module):
+    def __init__(
+            self,
+            output_size=512,
+            similarity_layers_sizes=[512, 512],
+            dropout=0.5,
+            output_type="classification",
+            n_classes=4,
+            freeze=False,
+    ):
+        """
+        Construct the Siamese network with ResNet as feature extractor
+        """
+        super().__init__()
+        self.left_network = get_pretrained_resnet34(output_size, freeze)
```

Comment:
Are the left and right networks trained separately? 
I see that the default value for freeze is False, thus both networks are being trained and you end up with different feature extractors. Is it how it's supposed to be?